requests-for-comments/2024-01-23-project-environments.md

```diff
@@ -0,0 +1,192 @@
+
+Extra internal requirement: Because Team Pipeline plans to implement multiple SDK keys per project, it would be simplest to move SDK keys to their own model, rather than moving to the environment model.
+
+<img width="939" alt="image" src="https://github.com/PostHog/meta/assets/4550621/ec966c6b-0a1c-4cb0-a593-0234f4807ab1">
+
+### Downsides
+
+Of course, nothing comes for free: adding evironments has its own downsides. Here, it's the amount of effort required – this feature would touch every part of the app, plus ingestion. It would realistically be a few person-sprints of work (as always, it's hard to estimate accurately).
+
+### Pricing
+
+Having multiple environments should start out as a Teams/Enterprise feature.
+
+### Rough scope of work
+
+1. Backend: We add an `Environment` model (`posthog_environment` table, columns int `id`, `team_id`, `data_id` `name`) and hook it up to a new project-scoped viewset (`/api/projects/:id/environments/`). Every time a new environment is created, we determine its `id` by incrementing the counter of the `posthog_team.id` column and using that value as the environment's `data_id`.
+1. Backend: For every project existing currently we create a matching default `Environment`, reusing the project's `id` the environment's `data_id`. Any time a new project is created, it also gets a default environment based on the same logic.
+1. Backend: We add an `SDKKey` model (columns UUID `id`, `environment_id`, `value`, `created_by`, `created_at`) and populate it with current `api_token` values from `posthog_team` (for each project using the project ID as `environment_id`).
+    > Until we replace the project API key in project settings with environments, we'll have to sync the project API key being reset with its `posthog_sdkkey` equivalent.
+1. Backend: We add environments to `TeamSerializer`.
+1. Backend: We add `current_environment` to the `User` model.
+1. Plugin server: We switch to `posthog_environment` as the source of truth for tokens and event/person `team_id`. From now on, `team_id` in ClickHouse will mean `posthog_environment.data_id`. The only thing we continue processing at the project level is event/property definitions.
+1. Frontend, behind a flag: We add logic for environment availability and selection to `teamLogic`.
+1. Frontend, behind a flag: We add the environment as a breadcrumb item with a dropdown + creation modal.
+    > We must make it crystal clear for customers splitting by region that cross-environment analytics isn't possible – environments are designed specifically to solve cross-contamination.
+1. Backend: To models of all optionally project-wide entities, we add boolean fields `environment_id` and `is_project_wide`:
+	- `Dashboard`
+	- `Insight`
+	- `Action`
+	- `Cohort`
+	- `FeatureFlag`
+	- `SessionRecordingPlaylist`
+1. Backend: In all endpoints that query ClickHouse, we add support for filtering by environment ID instead of project ID for querying:
```

Comment:
I added a comment further up about how we could re-use Teams for this instead. If we do that then this comment can basically be ignored, otherwise we should probably figure out this stuff as well as what do about frontend routing (how to link someone to a specific environment)